text/0000-expand-impl-trait.md

```diff
@@ -0,0 +1,853 @@
+
+- It provides a greater degree of analogy between static and dynamic dispatch
+  when working with traits. Introducing trait objects is easier when they can be
+  understood as a variant of `impl Trait`, rather than a completely different
+  approach. This effect would be further heightened if we moved to `dyn Trait`
+  syntax for trait objects.
+
+- It provides a more intuitive way of working with traits and static dispatch in
+  an "object" style, smoothing the transition to Rust's take on the topic.
+
+- It provides a more uniform story for static dispatch, allowing it to work in
+  both argument and return position.
+
+There are two ways of teaching `impl Trait`:
+
+- Introduce it *prior* to bounded generics, as the first way you learn to
+  "consume" traits. That works particularly well with teaching `Iterator` as one
+  of the first real traits you see, since `impl Trait` is a strong match for
+  working with iterators. As mentioned above, this approach also provides a more
+  intuitive stepping stone for those coming from OO-ish languages. Later,
+  bounded generics can be introduced as a more powerful, explicit syntax, which
+  can also reveal a bit more about the underlying semantic model of `impl
+  Trait`.  In this approach, the existential use case doesn't need a great deal
+  of ceremony---it just follows naturally from the basic feature.
+
+- Alternatively, introduce it *after* bounded generics, as (1) a sugar for
+  generics and (2) a separate mechanism for existentials. This is, of course,
+  the way all existing Rust users will come to learn `impl Trait`. And it's
+  ultimately important to understand the mechanism in this way. But it's likely
+  *not* the ideal way to introduce it at first.
+
+In either case, people should learn `impl Trait` early (since it will appear
+often) and in particular prior to learning trait objects. As mentioned above,
+trait objects can then be taught using intuitions from `impl Trait`.
+
+There's also some ways in which `impl Trait` can introduce confusion, which
+we'll cover in the drawbacks section below.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+It's widely recognized that we need *some* form of static existentials for
+return position, both to be able to return closures (which have un-nameable
+types) and to ergonomically return things like iterator chains.
+
+However, there are two broad classes of drawbacks to the approach taken in this RFC.
+
+## Relatively inexpressive sugary syntax
+
+This RFC is built on the idea that we'll eventually have a fully expressive
+explicit syntax, and so we should tailor the "sugary" `impl Trait` syntax to
+the most common use cases and intuitions.
+
+That means, however, that we give up an opportunity to provide more expressive
+but still sugary syntax like `some Trait` and `any Trait`---we certainly don't
+want all three.
+
+That syntax is further discussed in Alternatives below.
+
+## Potential for confusion
```

Comment:
This drawback really gets to me. I feel like it spends some complexity budget when there is more than one way to do a thing. However, I do feel like strong discouragement of usage of impl in an argument position might be sufficient. This kind of complexity impacts newcomers mostly, where they have to learn both ways to write a thing so that they can read code written by others. If we strongly discourage this, it is less of a problem. And then we get the benefit of impl trait working consistently-ish, which is nice and reduces confusion when learning impl trait.

Another issue here is that now you can make generic methods that don't look generic, because they don't have a `<T>` . This concerns me more. This concern is conceptually similar to the concern against eliding lifetimes in struct defs, though I personally am largely okay with that specific case.